common/src/main/java/io/crate/types/UncheckedObjectType.java

```diff
@@ -0,0 +1,115 @@
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import io.crate.common.collections.MapComparator;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Object type that makes no assumptions about neither the keys or values, treating them like generic values and lifting
+ * the restriction of having the keys be Strings.
+ */
+public class UncheckedObjectType extends DataType<Map<Object, Object>> implements Streamer<Map<Object, Object>> {
+
+    public static final UncheckedObjectType INSTANCE = new UncheckedObjectType();
+    public static final int ID = 16;
+
+    public static final String NAME = "unchecked_object";
+
+    public static UncheckedObjectType untyped() {
+        return new UncheckedObjectType();
+    }
+
+    UncheckedObjectType() {
+    }
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public Precedence precedence() {
+        return Precedence.UncheckedObjectType;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public Streamer<Map<Object, Object>> streamer() {
+        return this;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Map<Object, Object> value(Object value) {
+        Map<Object, Object> map = (Map<Object, Object>) value;
+        if (map == null) {
+            return map;
+        }
+        return new HashMap<>(map);
+    }
+
+    @Override
+    public int compareValueTo(Map<Object, Object> val1, Map<Object, Object> val2) {
+        return MapComparator.compareMaps(val1, val2);
+    }
+
+    @Override
+    public Map<Object, Object> readValueFrom(StreamInput in) throws IOException {
+        if (in.readBoolean()) {
+            int size = in.readInt();
+            HashMap<Object, Object> m = new HashMap<>(size);
+            for (int i = 0; i < size; i++) {
+                Object key = UndefinedType.INSTANCE.streamer().readValueFrom(in);
```

Comment:
Could we make `read` and `write` symmetric so that this either uses `in.readGenericValue()` or that `write` uses `UndefinedType.INSTANCE.streamer().writeValueTo(...)` ?